README
################################################################################
# color_diff.js
################################################################################
Implemets the CIEDE2000 color difference algorithm, conversion between RGB
and lab color and mapping all colors in palette X to the closest color in
palette Y based on the CIEDE2000 difference.

Tests:
There are some tests in color_diff_test.js, color_palette_test and
color_convert_test.js, run with:
node color_diff_test.js
node color_convert_test.js
node color_palette_test.js

Example usage:
> var rgbcolor1     = {'R' : 255, 'G': 255, 'B': 255});
> var rgbcolor2     = {'R' : 0, 'G': 0, 'B': 0});
> var labcolor1     = color_convert.rgb_to_lab(rgbcolor1);
> var labcolor2     = color_convert.rgb_to_lab(rgbcolor2);
> var diff          = color_diff.ciede2000(labcolor1, labcolor2);
> var palette_map   = color_palette.map_palette([rgbcolor1, rgbcolor2], [rgbcolor1]);
> var rgbcolor2key  = color_palette.palette_map_key(rgbcolor2)
> var rgbcolor2prim = palette_map[rgbcolor2key];
> rgbcolor1 == rgbcolor2prim
true

Author:
Markus Näsman

License:
See COPYING


CHANGELOG.md
# Changelog

## 1.4.0

* Use ESM modules. Backwards compatible with CJS using rollup.
* Replaced nyc with c8 (nyc does not support esm).

## 1.3.0

* don't support mixing casing for r,g,b properties, either all upper or all lower
* use camelCase instead of snake_case
* snake_case -> camelCase mappings are still present to not break more dependents than necessary.
* fixed jsdoc types and type checking using tsc
* only run tests on node > 18
* no more travis, use github actions
* bumped all deps
* added linting


test.sh
#!/bin/sh
set -e
for filename in test/*.js; do node "${filename}"; done



test/diff.js
/**
 * @author Markus Ekholm
 * @copyright 2012-2023 (c) Markus Ekholm <markus at botten dot org >
 * @license Copyright (c) 2012-2023, Markus Ekholm
 * All rights reserved.
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *    * Redistributions of source code must retain the above copyright
 *      notice, this list of conditions and the following disclaimer.
 *    * Redistributions in binary form must reproduce the above copyright
 *      notice, this list of conditions and the following disclaimer in the
 *      documentation and/or other materials provided with the distribution.
 *    * Neither the name of the author nor the
 *      names of its contributors may be used to endorse or promote products
 *      derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL MARKUS EKHOLM BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/**
 * IMPORTS
 */

import * as assert from "assert";

import * as colorDiff from "../lib/diff.js";

/**
 * TESTS
 */

/**
 * CIEDE2000 tests the diff.cie2000.
 * Cases taken from the paper "The CIEDE2000 Color-Difference Formula:
 * Implementation Notes, Supplementary Test Data, and Mathematical Observations"
 * by Gaurav Sharma, Wencheng Wu and Edul N. Dalal.
 */
describe("diff", () => {
  describe("#ciede2000()", () => {
    it("should use the true chroma difference (#1)", () => {
      assertCiede2000Diff(2.0425,
        {
          L: 50.0000,
          a: 2.6772,
          b: -79.7751,
        },
        {
          L: 50.0000,
          a: 0.0000,
          b: -82.7485,
        });
    });
    it("should use the true chroma difference (#2)", () => {
      assertCiede2000Diff(2.8615,
        {
          L: 50.0000,
          a: 3.1571,
          b: -77.2803,
        },
        {
          L: 50.0000,
          a: 0.0000,
          b: -82.7485,
        });
    });
    it("should use the true chroma difference (#3)", () => {
      assertCiede2000Diff(3.4412,
        {
          L: 50.0000,
          a: 2.8361,
          b: -74.0200,
        },
        {
          L: 50.0000,
          a: 0.0000,
          b: -82.7485,
        });
    });
    it("should use the true hue difference (#4)", () => {
      assertCiede2000Diff(1.0000,
        {
          L: 50.0000,
          a: -1.3802,
          b: -84.2814,
        },
        {
          L: 50.0000,
          a: 0.0000,
          b: -82.7485,
        });
    });
    it("should use the true hue difference (#5)", () => {
      assertCiede2000Diff(1.0000,
        {
          L: 50.0000,
          a: -1.1848,
          b: -84.8006,
        },
        {
          L: 50.0000,
          a: 0.0000,
          b: -82.7485,
        });
    });
    it("should use the true hue difference (#6)", () => {
      assertCiede2000Diff(1.0000,
        {
          L: 50.0000,
          a: -0.9009,
          b: -85.5211,
        },
        {
          L: 50.0000,
          a: 0.0000,
          b: -82.7485,
        });
    });
    it("should use the correct arctangent computation (#7)", () => {
      assertCiede2000Diff(2.3669,
        {
          L: 50.0000,
          a: 0.0000,
          b: 0.0000,
        },
        {
          L: 50.0000,
          a: -1.0000,
          b: 2.0000,
        });
    });
    it("should use the correct arctangent computation (#8)", () => {
      assertCiede2000Diff(2.3669,
        {
          L: 50.0000,
          a: -1.0000,
          b: 2.0000,
        },
        {
          L: 50.0000,
          a: 0.0000,
          b: 0.0000,
        });
    });
    it("should use the correct arctangent computation (#9)", () => {
      assertCiede2000Diff(7.1792,
        {
          L: 50.0000,
          a: 2.4900,
          b: -0.0010,
        },
        {
          L: 50.0000,
          a: -2.4900,
          b: 0.0009,
        });
    });
    it("should use the correct arctangent computation (#10)", () => {
      assertCiede2000Diff(7.1792,
        {
          L: 50.0000,
          a: 2.4900,
          b: -0.0010,
        },
        {
          L: 50.0000,
          a: -2.4900,
          b: 0.0010,
        });
    });
    it("should use the correct arctangent computation (#11)", () => {
      assertCiede2000Diff(7.2195,
        {
          L: 50.0000,
          a: 2.4900,
          b: -0.0010,
        },
        {
          L: 50.0000,
          a: -2.4900,
          b: 0.0011,
        });
    });
    it("should use the correct arctangent computation (#12)", () => {
      assertCiede2000Diff(7.2195,
        {
          L: 50.0000,
          a: 2.4900,
          b: -0.0010,
        },
        {
          L: 50.0000,
          a: -2.4900,
          b: 0.0012,
        });
    });
    it("should use the correct arctangent computation (#13)", () => {
      assertCiede2000Diff(4.8045,
        {
          L: 50.0000,
          a: -0.0010,
          b: 2.4900,
        },
        {
          L: 50.0000,
          a: 0.0009,
          b: -2.4900,
        });
    });
    it("should use the correct arctangent computation (#14)", () => {
      assertCiede2000Diff(4.8045,
        {
          L: 50.0000,
          a: -0.0010,
          b: 2.4900,
        },
        {
          L: 50.0000,
          a: 0.0010,
          b: -2.4900,
        });
    });
    it("should use the correct arctangent computation (#15)", () => {
      assertCiede2000Diff(4.7461,
        {
          L: 50.0000,
          a: -0.0010,
          b: 2.4900,
        },
        {
          L: 50.0000,
          a: 0.0011,
          b: -2.4900,
        });
    });
    it("should use the correct arctangent computation (#16)", () => {
      assertCiede2000Diff(4.3065,
        {
          L: 50.0000,
          a: 2.5000,
          b: 0.0000,
        },
        {
          L: 50.0000,
          a: 0.0000,
          b: -2.5000,
        });
    });
    it("should work for large color differences (#17)", () => {
      assertCiede2000Diff(27.1492,
        {
          L: 50.0000,
          a: 2.5000,
          b: 0.0000,
        },
        {
          L: 73.0000,
          a: 25.0000,
          b: -18.0000,
        });
    });
    it("should work for large color differences (#18)", () => {
      assertCiede2000Diff(22.8977,
        {
          L: 50.0000,
          a: 2.5000,
          b: 0.0000,
        },
        {
          L: 61.0000,
          a: -5.0000,
          b: 29.0000,
        });
    });
    it("should work for large color differences (#19)", () => {
      assertCiede2000Diff(31.9030,
        {
          L: 50.0000,
          a: 2.5000,
          b: 0.0000,
        },
        {
          L: 56.0000,
          a: -27.0000,
          b: -3.0000,
        });
    });
    it("should work for large color differences (#20)", () => {
      assertCiede2000Diff(19.4535,
        {
          L: 50.0000,
          a: 2.5000,
          b: 0.0000,
        },
        {
          L: 58.0000,
          a: 24.0000,
          b: 15.0000,
        });
    });
    it("should produce numbers found in the CIE technical report (#21)", () => {
      assertCiede2000Diff(1.0000,
        {
          L: 50.0000,
          a: 2.5000,
          b: 0.0000,
        },
        {
          L: 50.0000,
          a: 3.1736,
          b: 0.5854,
        });
    });
    it("should produce numbers found in the CIE technical report (#22)", () => {
      assertCiede2000Diff(1.0000,
        {
          L: 50.0000,
          a: 2.5000,
          b: 0.0000,
        },
        {
          L: 50.0000,
          a: 3.2972,
          b: 0.0000,
        });
    });
    it("should produce numbers found in the CIE technical report (#23)", () => {
      assertCiede2000Diff(1.0000,
        {
          L: 50.0000,
          a: 2.5000,
          b: 0.0000,
        },
        {
          L: 50.0000,
          a: 1.8634,
          b: 0.5757,
        });
    });
    it("should produce numbers found in the CIE technical report (#24)", () => {
      assertCiede2000Diff(1.0000,
        {
          L: 50.0000,
          a: 2.5000,
          b: 0.0000,
        },
        {
          L: 50.0000,
          a: 3.2592,
          b: 0.3350,
        });
    });
    it("should produce numbers found in the CIE technical report (#25)", () => {
      assertCiede2000Diff(1.2644,
        {
          L: 60.2574,
          a: -34.0099,
          b: 36.2677,
        },
        {
          L: 60.4626,
          a: -34.1751,
          b: 39.4387,
        });
    });
    it("should produce numbers found in the CIE technical report (#26)", () => {
      assertCiede2000Diff(1.2630,
        {
          L: 63.0109,
          a: -31.0961,
          b: -5.8663,
        },
        {
          L: 62.8187,
          a: -29.7946,
          b: -4.0864,
        });
    });
    it("should produce numbers found in the CIE technical report (#27)", () => {
      assertCiede2000Diff(1.8731,
        {
          L: 61.2901,
          a: 3.7196,
          b: -5.3901,
        },
        {
          L: 61.4292,
          a: 2.2480,
          b: -4.9620,
        });
    });
    it("should produce numbers found in the CIE technical report (#28)", () => {
      assertCiede2000Diff(1.8645,
        {
          L: 35.0831,
          a: -44.1164,
          b: 3.7933,
        },
        {
          L: 35.0232,
          a: -40.0716,
          b: 1.5901,
        });
    });
    it("should produce numbers found in the CIE technical report (#29)", () => {
      assertCiede2000Diff(2.0373,
        {
          L: 22.7233,
          a: 20.0904,
          b: -46.6940,
        },
        {
          L: 23.0331,
          a: 14.9730,
          b: -42.5619,
        });
    });
    it("should produce numbers found in the CIE technical report (#30)", () => {
      assertCiede2000Diff(1.4146,
        {
          L: 36.4612,
          a: 47.8580,
          b: 18.3852,
        },
        {
          L: 36.2715,
          a: 50.5065,
          b: 21.2231,
        });
    });
    it("should produce numbers found in the CIE technical report (#31)", () => {
      assertCiede2000Diff(1.4441,
        {
          L: 90.8027,
          a: -2.0831,
          b: 1.4410,
        },
        {
          L: 91.1528,
          a: -1.6435,
          b: 0.0447,
        });
    });
    it("should produce numbers found in the CIE technical report (#32)", () => {
      assertCiede2000Diff(1.5381,
        {
          L: 90.9257,
          a: -0.5406,
          b: -0.9208,
        },
        {
          L: 88.6381,
          a: -0.8985,
          b: -0.7239,
        });
    });
    it("should produce numbers found in the CIE technical report (#33)", () => {
      assertCiede2000Diff(0.6377,
        {
          L: 6.7747,
          a: -0.2908,
          b: -2.4247,
        },
        {
          L: 5.8714,
          a: -0.0985,
          b: -2.2286,
        });
    });
    it("should produce numbers found in the CIE technical report (#34)", () => {
      assertCiede2000Diff(0.9082,
        {
          L: 2.0776,
          a: 0.0795,
          b: -1.1350,
        },
        {
          L: 0.9033,
          a: -0.0636,
          b: -0.5514,
        });
    });
    it("Same color should have 0.0 difference #1", () => {
      assertCiede2000Diff(0.0,
        { L: 100, a: 0.005, b: -0.010 },
        { L: 100, a: 0.005, b: -0.010 }
      );
    });
    it("Same color should have 0.0 difference #2", () => {
      assertCiede2000Diff(0.0,
        { L: 0.0, a: 0.0, b: 0.0 },
        { L: 0.0, a: 0.0, b: 0.0 }
      );
    });
    it("Black and white are very different", () => {
      assertCiede2000Diff(100.0,
        { L: 100, a: 0.005, b: -0.010 },
        { L: 0.0, a: 0.0, b: 0.0 }
      );
    });
    it("should throw error", () => {
      assert.throws(
        () => {
          colorDiff.ciede2000({
            L: NaN,
            a: NaN,
            b: NaN,
          },
          {
            L: 0,
            a: 0,
            b: 0,
          });
        },
        Error);
    });
  });
});

/**
 * INTERNAL FUNCTIONS
 */

/**
 *
 * @param {number} expected
 * @param {import("../").LabColor} c1
 * @param {import("../").LabColor} c2
 */
function assertCiede2000Diff(expected, c1, c2) {
  assert.equal(expected, round(colorDiff.ciede2000(c1, c2)));
  assert.equal(expected, round(colorDiff.ciede2000(c2, c1)));
}

/**
 *
 * @param {number} n
 * @returns number
 */
function round(n) {
  return Math.round(n * 10000) / 10000;
}


test/index.js
/**
 * IMPORTS
 */
var assert        = require('assert');
var colorDiff     = require('../lib/index');
var color_convert = require('../lib/convert');

/**
 * CONSTANTS
 */

var pad = function (s) { return s.length === 1 ? "0" + s : s; };
var hex = function (n) { return pad(n.toString(16)); }
var rgb_to_hex = function (c) { return "#" + hex(c.R) + hex(c.G) + hex(c.B); };

var white   = {'R':255 , 'G':255 ,'B':255};
var black   = {'R':0   , 'G':0   ,'B':0};
var navy    = {'R':0   , 'G':0   ,'B':128};
var blue    = {'R':0   , 'G':0   ,'B':255};
var yellow  = {'R':255 , 'G':255 ,'B':0};
var gold    = {'R':255 , 'G':215 ,'B':0};

var white_hex   = rgb_to_hex({'R':255 , 'G':255 ,'B':255});
var black_hex   = rgb_to_hex({'R':0   , 'G':0   ,'B':0});
var navy_hex    = rgb_to_hex({'R':0   , 'G':0   ,'B':128});
var blue_hex    = rgb_to_hex({'R':0   , 'G':0   ,'B':255});
var yellow_hex  = rgb_to_hex({'R':255 , 'G':255 ,'B':0});
var gold_hex    = rgb_to_hex({'R':255 , 'G':215 ,'B':0});

var colors1 = [white, black, navy, blue, yellow, gold]
var colors2 = [white, black, blue, gold]
var colors3 = [white, black, yellow, blue]

var colors1_hex = [white_hex, black_hex, navy_hex, blue_hex, yellow_hex, gold_hex]
var colors2_hex = [white_hex, black_hex, blue_hex, gold_hex]
var colors3_hex = [white_hex, black_hex, yellow_hex, blue_hex]

var colors1_hex_prepared = colorDiff.prepare_palette(colors1_hex);
var colors2_hex_prepared = colorDiff.prepare_palette(colors2_hex);
var colors3_hex_prepared = colorDiff.prepare_palette(colors3_hex);

/**
 * TESTS
 */

describe('index', function(){
  describe('#prepare_palette', function (){
    it('should convert hex through to Lab',
      function() {
        var original = ["#fFfFfF", "#abcdef", "#000000"];
        var prepared = colorDiff.prepare_palette(original);
        assert.deepEqual(prepared, {
          prepared: true,
          original: original,
          lab: [
            color_convert.rgb_to_lab({R: 255, G: 255, B: 255}),
            color_convert.rgb_to_lab({R: 171, G: 205, B: 239}),
            color_convert.rgb_to_lab({R: 0, G: 0, B: 0}),
          ]
        });
      });
  });

  describe('process rgb', function (){
    it('should find correct closest colors',
      function() {
         assert.deepEqual(colorDiff.closest(white, colors1), white);
         assert.deepEqual(colorDiff.closest(black, colors1), black);
         assert.deepEqual(colorDiff.closest(navy, colors1), navy);
         assert.deepEqual(colorDiff.closest(blue, colors1), blue);
         assert.deepEqual(colorDiff.closest(yellow, colors1), yellow);
         assert.deepEqual(colorDiff.closest(gold, colors1), gold);

         assert.deepEqual(colorDiff.closest(white, colors2), white);
         assert.deepEqual(colorDiff.closest(black, colors2), black);
         assert.deepEqual(colorDiff.closest(navy, colors2), blue);
         assert.deepEqual(colorDiff.closest(blue, colors2), blue);
         assert.deepEqual(colorDiff.closest(yellow, colors2), gold);
         assert.deepEqual(colorDiff.closest(gold, colors2), gold);
      });
    it('should find correct furthest colors',
      function() {
         assert.deepEqual(colorDiff.furthest(white, colors3), black);
         assert.deepEqual(colorDiff.furthest(black, colors3), yellow);
         assert.deepEqual(colorDiff.furthest(navy, colors3), yellow);
         assert.deepEqual(colorDiff.furthest(blue, colors3), yellow);
         assert.deepEqual(colorDiff.furthest(yellow, colors3), blue);
         assert.deepEqual(colorDiff.furthest(gold, colors3), blue);
      });
  });

  describe('process hex', function (){
    it('should find correct closest colors',
      function() {
         assert.deepEqual(colorDiff.closest(white_hex, colors1_hex), white_hex);
         assert.deepEqual(colorDiff.closest(black_hex, colors1_hex), black_hex);
         assert.deepEqual(colorDiff.closest(navy_hex, colors1_hex), navy_hex);
         assert.deepEqual(colorDiff.closest(blue_hex, colors1_hex), blue_hex);
         assert.deepEqual(colorDiff.closest(yellow_hex, colors1_hex), yellow_hex);
         assert.deepEqual(colorDiff.closest(gold_hex, colors1_hex), gold_hex);

         assert.deepEqual(colorDiff.closest(white_hex, colors2_hex), white_hex);
         assert.deepEqual(colorDiff.closest(black_hex, colors2_hex), black_hex);
         assert.deepEqual(colorDiff.closest(navy_hex, colors2_hex), blue_hex);
         assert.deepEqual(colorDiff.closest(blue_hex, colors2_hex), blue_hex);
         assert.deepEqual(colorDiff.closest(yellow_hex, colors2_hex), gold_hex);
         assert.deepEqual(colorDiff.closest(gold_hex, colors2_hex), gold_hex);
      });

    it('should find correct furthest colors',
      function() {
         assert.deepEqual(colorDiff.furthest(white_hex, colors3_hex), black_hex);
         assert.deepEqual(colorDiff.furthest(black_hex, colors3_hex), yellow_hex);
         assert.deepEqual(colorDiff.furthest(navy_hex, colors3_hex), yellow_hex);
         assert.deepEqual(colorDiff.furthest(blue_hex, colors3_hex), yellow_hex);
         assert.deepEqual(colorDiff.furthest(yellow_hex, colors3_hex), blue_hex);
         assert.deepEqual(colorDiff.furthest(gold_hex, colors3_hex), blue_hex);
      });
  })

  describe('process hex prepared palette', function (){
    it('should find correct closest colors',
      function() {
         assert.deepEqual(colorDiff.closest(white_hex, colors1_hex_prepared), white_hex);
         assert.deepEqual(colorDiff.closest(black_hex, colors1_hex_prepared), black_hex);
         assert.deepEqual(colorDiff.closest(navy_hex, colors1_hex_prepared), navy_hex);
         assert.deepEqual(colorDiff.closest(blue_hex, colors1_hex_prepared), blue_hex);
         assert.deepEqual(colorDiff.closest(yellow_hex, colors1_hex_prepared), yellow_hex);
         assert.deepEqual(colorDiff.closest(gold_hex, colors1_hex_prepared), gold_hex);

         assert.deepEqual(colorDiff.closest(white_hex, colors2_hex_prepared), white_hex);
         assert.deepEqual(colorDiff.closest(black_hex, colors2_hex_prepared), black_hex);
         assert.deepEqual(colorDiff.closest(navy_hex, colors2_hex_prepared), blue_hex);
         assert.deepEqual(colorDiff.closest(blue_hex, colors2_hex_prepared), blue_hex);
         assert.deepEqual(colorDiff.closest(yellow_hex, colors2_hex_prepared), gold_hex);
         assert.deepEqual(colorDiff.closest(gold_hex, colors2_hex_prepared), gold_hex);
      });

    it('should find correct furthest colors',
      function() {
         assert.deepEqual(colorDiff.furthest(white_hex, colors3_hex_prepared), black_hex);
         assert.deepEqual(colorDiff.furthest(black_hex, colors3_hex_prepared), yellow_hex);
         assert.deepEqual(colorDiff.furthest(navy_hex, colors3_hex_prepared), yellow_hex);
         assert.deepEqual(colorDiff.furthest(blue_hex, colors3_hex_prepared), yellow_hex);
         assert.deepEqual(colorDiff.furthest(yellow_hex, colors3_hex_prepared), blue_hex);
         assert.deepEqual(colorDiff.furthest(gold_hex, colors3_hex_prepared), blue_hex);
      });
  })
});

// Local Variables:
// allout-layout: t
// js-indent-level: 2
// End:


color_convert.js
/**
 * @name color_convert.js
 * @author Markus Näsman
 * @copyright 2012 (c) Markus Näsman <markus at botten dot org >
 * @license see COPYING
 */

/**
* EXPORTS
*/
exports.rgb_to_lab = rgb_to_lab;

/**
* IMPORTS
*/
var pow  = Math.pow;
var sqrt = Math.sqrt;

/**
 * API FUNCTIONS
 */

/**
* Returns c converted to labcolor.
* @param {rgbcolor} c should have fields R,G,B
* @return {labcolor} c converted to labcolor
*/
function rgb_to_lab(c)
{
  return xyz_to_lab(rgb_to_xyz(c))
}

/**
* Returns c converted to xyzcolor.
* @param {rgbcolor} c should have fields R,G,B
* @return {xyzcolor} c converted to xyzcolor
*/
function rgb_to_xyz(c)
{
  // Based on http://www.easyrgb.com/index.php?X=MATH&H=02
  var R = ( c.R / 255 );
  var G = ( c.G / 255 );
  var B = ( c.B / 255 );

  if ( R > 0.04045 ) R = pow(( ( R + 0.055 ) / 1.055 ),2.4);
  else               R = R / 12.92;
  if ( G > 0.04045 ) G = pow(( ( G + 0.055 ) / 1.055 ),2.4);
  else               G = G / 12.92;
  if ( B > 0.04045 ) B = pow(( ( B + 0.055 ) / 1.055 ), 2.4);
  else               B = B / 12.92;

  R *= 100;
  G *= 100;
  B *= 100;

  // Observer. = 2°, Illuminant = D65
  var X = R * 0.4124 + G * 0.3576 + B * 0.1805;
  var Y = R * 0.2126 + G * 0.7152 + B * 0.0722;
  var Z = R * 0.0193 + G * 0.1192 + B * 0.9505;
  return {'X' : X, 'Y' : Y, 'Z' : Z};
}

/**
* Returns c converted to labcolor.
* @param {xyzcolor} c should have fields X,Y,Z
* @return {labcolor} c converted to labcolor
*/
function xyz_to_lab(c)
{
  var ref_Y = 100.000;
  var ref_Z = 108.883;
  var ref_X = 95.047; // Observer= 2°, Illuminant= D65
  var Y = c.Y / ref_Y;
  var Z = c.Z / ref_Z;
  var X = c.X / ref_X;
  if ( X > 0.008856 ) X = pow(X, 1/3);
  else                X = ( 7.787 * X ) + ( 16 / 116 );
  if ( Y > 0.008856 ) Y = pow(Y, 1/3);
  else                Y = ( 7.787 * Y ) + ( 16 / 116 );
  if ( Z > 0.008856 ) Z = pow(Z, 1/3);
  else                Z = ( 7.787 * Z ) + ( 16 / 116 );
  var L = ( 116 * Y ) - 16;
  var a = 500 * ( X - Y );
  var b = 200 * ( Y - Z );
  return {'L' : L , 'a' : a, 'b' : b};
}

// Local Variables:
// allout-layout: t
// js-indent-level: 2
// End:


lib/diff.js
/**
 * @author Markus Ekholm
 * @copyright 2012-2023 (c) Markus Ekholm <markus at botten dot org >
 * @license Copyright (c) 2012-2023, Markus Ekholm
 * All rights reserved.
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *    * Redistributions of source code must retain the above copyright
 *      notice, this list of conditions and the following disclaimer.
 *    * Redistributions in binary form must reproduce the above copyright
 *      notice, this list of conditions and the following disclaimer in the
 *      documentation and/or other materials provided with the distribution.
 *    * Neither the name of the author nor the
 *      names of its contributors may be used to endorse or promote products
 *      derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL MARKUS EKHOLM BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/**
* IMPORTS
*/
import { rgbaToLab } from "./convert.js";

/**
* EXPORTS
*/
export { ciede2000 };

/**
 * TYPES
 */

/**
 * @typedef {import("../").LabColor} LabColor
 * @typedef {import("../").RGBAColor} RGBAColor
 * @typedef {import("../").Color} Color
 */

/**
 * API FUNCTIONS
 */

/**
* Returns diff between c1 and c2 using the CIEDE2000 algorithm
* @param {Color} c1
* @param {Color} c2
* @param {RGBAColor} [bc] background color
* @return {number} Difference between c1 and c2
*/
function ciede2000(c1, c2, bc) {
  if ("R" in c1 || "r" in c1) {
    c1 = rgbaToLab(c1, bc);
  }

  if ("R" in c2 || "r" in c2) {
    c2 = rgbaToLab(c2, bc);
  }
  /**
   * Implemented as in "The CIEDE2000 Color-Difference Formula:
   * Implementation Notes, Supplementary Test Data, and Mathematical Observations"
   * by Gaurav Sharma, Wencheng Wu and Edul N. Dalal.
   */

  // Get L,a,b values for color 1
  const L1 = c1.L;
  const a1 = c1.a;
  const b1 = c1.b;

  // Get L,a,b values for color 2
  const L2 = c2.L;
  const a2 = c2.a;
  const b2 = c2.b;

  // Weight factors
  const kL = 1;
  const kC = 1;
  const kH = 1;

  /**
   * Step 1: Calculate C1p, C2p, h1p, h2p
   */
  const C1 = Math.sqrt(Math.pow(a1, 2) + Math.pow(b1, 2)); // (2)
  const C2 = Math.sqrt(Math.pow(a2, 2) + Math.pow(b2, 2)); // (2)

  const aC1C2 = (C1 + C2) / 2.0; // (3)

  const G = 0.5 * (1 - Math.sqrt(Math.pow(aC1C2, 7.0) /
                          (Math.pow(aC1C2, 7.0) + Math.pow(25.0, 7.0)))); // (4)

  const a1p = (1.0 + G) * a1; // (5)
  const a2p = (1.0 + G) * a2; // (5)

  const C1p = Math.sqrt(Math.pow(a1p, 2) + Math.pow(b1, 2)); // (6)
  const C2p = Math.sqrt(Math.pow(a2p, 2) + Math.pow(b2, 2)); // (6)

  const h1p = hpF(b1, a1p); // (7)
  const h2p = hpF(b2, a2p); // (7)

  /**
   * Step 2: Calculate dLp, dCp, dHp
   */
  const dLp = L2 - L1; // (8)
  const dCp = C2p - C1p; // (9)

  const dhp = dhpF(C1, C2, h1p, h2p); // (10)
  const dHp = 2 * Math.sqrt(C1p * C2p) * Math.sin(radians(dhp) / 2.0); // (11)

  /**
   * Step 3: Calculate CIEDE2000 Color-Difference
   */
  const aL = (L1 + L2) / 2.0; // (12)
  const aCp = (C1p + C2p) / 2.0; // (13)

  const aHp = aHpF(C1, C2, h1p, h2p); // (14)
  const T = 1 - 0.17 * Math.cos(radians(aHp - 30)) + 0.24 * Math.cos(radians(2 * aHp)) +
    0.32 * Math.cos(radians(3 * aHp + 6)) - 0.20 * Math.cos(radians(4 * aHp - 63)); // (15)
  const dRo = 30 * Math.exp(-(Math.pow((aHp - 275) / 25, 2))); // (16)
  const RC = Math.sqrt((Math.pow(aCp, 7.0)) / (Math.pow(aCp, 7.0) + Math.pow(25.0, 7.0)));// (17)
  const SL = 1 + ((0.015 * Math.pow(aL - 50, 2)) /
                Math.sqrt(20 + Math.pow(aL - 50, 2.0)));// (18)
  const SC = 1 + 0.045 * aCp;// (19)
  const SH = 1 + 0.015 * aCp * T;// (20)
  const RT = -2 * RC * Math.sin(radians(2 * dRo));// (21)
  const dE = Math.sqrt(Math.pow(dLp / (SL * kL), 2) + Math.pow(dCp / (SC * kC), 2) +
                Math.pow(dHp / (SH * kH), 2) + RT * (dCp / (SC * kC)) *
                (dHp / (SH * kH))); // (22)
  return dE;
}

/**
 * INTERNAL FUNCTIONS
 */

/**
 *
 * @param {number} n
 * @returns {number}
 */
function degrees(n) {
  return n * (180 / Math.PI);
}

/**
 *
 * @param {number} n
 * @returns number
 */
function radians(n) {
  return n * (Math.PI / 180);
}

/**
 *
 * @param {number} x
 * @param {number} y
 * @returns {number}
 */
function hpF(x, y) { // (7)
  if (x === 0 && y === 0) return 0;
  else {
    const tmphp = degrees(Math.atan2(x, y));
    if (tmphp >= 0) return tmphp;
    else return tmphp + 360;
  }
}

/**
 *
 * @param {number} C1
 * @param {number} C2
 * @param {number} h1p
 * @param {number} h2p
 * @returns {number}
 */
function dhpF(C1, C2, h1p, h2p) { // (10)
  if (C1 * C2 === 0) return 0;
  else if (Math.abs(h2p - h1p) <= 180) return h2p - h1p;
  else if ((h2p - h1p) > 180) return (h2p - h1p) - 360;
  else if ((h2p - h1p) < -180) return (h2p - h1p) + 360;
  else throw (new Error());
}

/**
 *
 * @param {number} C1
 * @param {number} C2
 * @param {number} h1p
 * @param {number} h2p
 * @returns {number}
 */
function aHpF(C1, C2, h1p, h2p) { // (14)
  if (C1 * C2 === 0) return h1p + h2p;
  else if (Math.abs(h1p - h2p) <= 180) return (h1p + h2p) / 2.0;
  else if ((Math.abs(h1p - h2p) > 180) && ((h1p + h2p) < 360)) return (h1p + h2p + 360) / 2.0;
  else if ((Math.abs(h1p - h2p) > 180) && ((h1p + h2p) >= 360)) return (h1p + h2p - 360) / 2.0;
  else throw (new Error());
}
